h/services/group.py

```diff
@@ -67,11 +65,25 @@ def create_open_group(self, name, userid, description=None):
         return self._create(name=name,
                             userid=userid,
                             description=description,
-                            joinable_by=None,
-                            readable_by=ReadableBy.world,
-                            writeable_by=WriteableBy.authority,
+                            access_flags=_OpenGroupMatcher,
                             )
 
+    def type(self, group):
```

Comment:
Hmm. The problem is that if the methods for _creating_ groups of different types are in `GroupService` and then the method for querying the type of a group is in the `Group` model, then the definitions of the different group types (which are currently in `_PrivateGroupMatcher` nd `_OpenGroupMatcher`) would be needed in _both_ `GroupService` and in `Group`.

I was thinking that `GroupService` would contain both the methods for creating groups of different types (`GroupService.create_private_group()` and `GroupService.create_open_group()`) and in future also the methods for getting all the groups of a given type (`GroupService.open_groups_for(url)`). And also `type()`. So then the definitions of the different group types are contained within `GroupService`, and  the `Group` model itself doesn't know about them.

But it is awkward to have to do `group_service.type(my_group)`.

Maybe the way to keep the definitions of the group types in one place is to push all usage of them down into the model then? So `Group` could have `@classmethod`s `create_open_group()` and `create_private_group()` and an `@property type`.

I think we'd still need `GroupService.create_open_group()` and `GroupService.create_private_group()` because these methods currently do some non-groups-model stuff as well, like publishing events, and creating `GroupScopes` for an open group, and adding members to private groups.

So something like:

```python
class Group:
    @classmethod
    def create_private_group(self, ...):
        ...

    @classmethod
    def create_open_group(self, ...):
        ...

    @property
    def type(self):
        ...

class GroupService:
    def create_private_group(self, ...):
       ... (calls Group.create_private_group() and does some other stuff as well)

    def create_open_group(self, ...):
       ... (calls Group.create_open_group() and does some other stuff as well)

    def open_groups_for(self, url):
      ... (Returns all open groups whose scope matches a given URL. Could go in a different service instead.)
```

What do you think?

With this approach we're spreading the concepts of the open and private group types throughout the model and at least `GroupService` and possibly other services too. But at least the actual definitions of what the different types mean (the access flags) are contained within the model class.

Reply:
Yeah, I think I like where this is going. Something intuitively suggests to me that we should be able to ask the `Group` about its type without an extra service layer, as you were alluding to. The service still absolutely makes sense otherwise, it's just that `type` notion...

Reply:
Yeah, I think `type` belongs on the group as well.

My intention was to keep types out of the groups model entirely, but maybe that needs to be reduced to just keeping them out of the database itself. So it's just a single `group` table in the db (not separate `opengroup` and `privategroup` tables, and no `group.type` column in the db either), and it's just a single ORM class `Group`, but the `Group` class does know all about the different types.